In [ ]:
!pip install tensorflow==2.15.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 475.3/475.3 MB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 32.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 41.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 63.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 442.0/442.0 kB 35.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.4/78.4 kB 7.7 MB/s eta 0:00:00
  Attempting uninstall: wrapt
    Found existing installation: wrapt 1.17.2
    Uninstalling wrapt-1.17.2:
      Successfully uninstalled wrapt-1.17.2
  Attempting uninstall: ml-dtypes
    Found existing installation: ml-dtypes 0.4.1
    Uninstalling ml-dtypes-0.4.1:
      Successfully uninstalled ml-dtypes-0.4.1
  Attempting uninstall: keras
    Found existing installation: keras 3.8.0
    Uninstalling keras-3.8.0:
      Successfully uninstalled keras-3.8.0
  Attempting uninstall: tensorboard
    Found existing installation

In [ ]:
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense
import time

In [ ]:
file = open(r"next_word_predictor.txt", "r", encoding = "utf8")
# Read the contents directly from the 'file' object
txt = file.read()

print(txt[:500])

The sun was shining brightly in the clear blue sky, and a gentle breeze rustled the leaves of the tall trees. People were out enjoying the beautiful weather, some sitting in the park, others taking a leisurely stroll along the riverbank. Children were playing games, and laughter filled the air.

As the day turned into evening, the temperature started to drop, and the sky transformed into a canvas of vibrant colors. Families gathered for picnics, and the smell of barbecues wafted through the air.


In [ ]:
# Assign Tokens to every word in the text

tokenizer =  Tokenizer()
tokenizer.fit_on_texts([txt])
tokenizer.word_index

{'the': 1,
 'and': 2,
 'a': 3,
 'of': 4,
 'to': 5,
 'i': 6,
 'you': 7,
 'in': 8,
 'is': 9,
 'monica': 10,
 'it': 11,
 'with': 12,
 'ross': 13,
 'that': 14,
 'rachel': 15,
 'for': 16,
 'chandler': 17,
 'this': 18,
 'on': 19,
 'joey': 20,
 'was': 21,
 'oh': 22,
 'phoebe': 23,
 'are': 24,
 'all': 25,
 'as': 26,
 'what': 27,
 'be': 28,
 'like': 29,
 'no': 30,
 "it's": 31,
 "i'm": 32,
 'her': 33,
 'they': 34,
 'just': 35,
 'from': 36,
 'okay': 37,
 'not': 38,
 'so': 39,
 'my': 40,
 'have': 41,
 'me': 42,
 'where': 43,
 'know': 44,
 'she': 45,
 'we': 46,
 'out': 47,
 'well': 48,
 'their': 49,
 'can': 50,
 'at': 51,
 'he': 52,
 'yeah': 53,
 'your': 54,
 'about': 55,
 'but': 56,
 'its': 57,
 'up': 58,
 "don't": 59,
 'text': 60,
 'scene': 61,
 'by': 62,
 'do': 63,
 'an': 64,
 'or': 65,
 'were': 66,
 'there': 67,
 'if': 68,
 'uh': 69,
 'look': 70,
 'life': 71,
 'through': 72,
 'into': 73,
 'him': 74,
 'his': 75,
 "you're": 76,
 'hey': 77,
 'how': 78,
 'right': 79,
 'think': 80,
 'time': 81,
 'no

Word Index Length

In [ ]:
word_index_length = len(tokenizer.word_index)
print(f"The length of the word index is : {word_index_length}")

The length of the word index is : 4993


Generating Input Sequences

The input_sequences list contains sequences of tokens generated from the cleaned data.

Each sequence consists of a variable number of tokens, representing the context and target words for training a language model.

The printed output shows the first few input sequences as a demonstration.

In [ ]:
 #Splitting text into sentences on the basis of '\n' and tokenize input sequence
 for sentence in txt.split('\n'):
  tokenized_sentence = tokenizer.texts_to_sequences([sentence])[0]
  print(tokenized_sentence)

[1, 155, 21, 2368, 1549, 8, 1, 422, 692, 215, 2, 3, 2369, 1550, 2370, 1, 423, 4, 1, 1142, 491, 84, 66, 47, 1143, 1, 693, 1144, 98, 575, 8, 1, 576, 492, 493, 3, 1145, 2371, 694, 1, 2372, 1146, 66, 494, 2373, 2, 368, 243, 1, 115]
[]
[26, 1, 244, 1147, 73, 1551, 1, 1552, 873, 5, 695, 2, 1, 215, 696, 73, 3, 874, 4, 99, 697, 1553, 875, 16, 2374, 2, 1, 876, 4, 2375, 1554, 72, 1, 115, 11, 21, 3, 698, 244, 16, 3, 2376, 62, 1, 2377]
[]
[8, 1, 1555, 7, 105, 369, 1, 424, 4, 275, 303, 699, 36, 3, 304, 2378, 2, 84, 1556, 5, 877, 190, 1, 2379, 5, 700, 1, 2380, 1, 577, 21, 1148, 2, 1, 303, 85, 151, 2381, 5, 1, 1557]
[]
[26, 1, 425, 1556, 5, 1558, 8, 1, 125, 215, 1, 2382, 1559, 191, 2383, 2, 1, 2384, 878, 48, 73, 1, 125, 11, 21, 3, 244, 243, 12, 1560, 368, 2, 2385, 14, 131, 169, 3, 2386]
[]
[]
[1, 94, 578, 1149, 19, 3, 1561, 57, 426, 1562, 2387, 58, 1150, 1, 215, 1, 578, 85, 3, 192, 145, 2, 57, 1151, 879, 85, 1563, 701, 1564, 2, 2388, 2389, 2390, 36, 25, 106, 1, 86, 2391, 5, 702, 57, 1152]
[]
[880, 1,

In [ ]:
# Convert Every sentence into input sequences
input_sequences = []
for sentence in txt.split('\n'):
  tokenized_sentence = tokenizer.texts_to_sequences([sentence])[0]

  for i in range(1, len(tokenized_sentence)):
    input_sequences.append(tokenized_sentence[:i+1])

In [ ]:
input_sequences

[[1, 155],
 [1, 155, 21],
 [1, 155, 21, 2368],
 [1, 155, 21, 2368, 1549],
 [1, 155, 21, 2368, 1549, 8],
 [1, 155, 21, 2368, 1549, 8, 1],
 [1, 155, 21, 2368, 1549, 8, 1, 422],
 [1, 155, 21, 2368, 1549, 8, 1, 422, 692],
 [1, 155, 21, 2368, 1549, 8, 1, 422, 692, 215],
 [1, 155, 21, 2368, 1549, 8, 1, 422, 692, 215, 2],
 [1, 155, 21, 2368, 1549, 8, 1, 422, 692, 215, 2, 3],
 [1, 155, 21, 2368, 1549, 8, 1, 422, 692, 215, 2, 3, 2369],
 [1, 155, 21, 2368, 1549, 8, 1, 422, 692, 215, 2, 3, 2369, 1550],
 [1, 155, 21, 2368, 1549, 8, 1, 422, 692, 215, 2, 3, 2369, 1550, 2370],
 [1, 155, 21, 2368, 1549, 8, 1, 422, 692, 215, 2, 3, 2369, 1550, 2370, 1],
 [1, 155, 21, 2368, 1549, 8, 1, 422, 692, 215, 2, 3, 2369, 1550, 2370, 1, 423],
 [1,
  155,
  21,
  2368,
  1549,
  8,
  1,
  422,
  692,
  215,
  2,
  3,
  2369,
  1550,
  2370,
  1,
  423,
  4],
 [1,
  155,
  21,
  2368,
  1549,
  8,
  1,
  422,
  692,
  215,
  2,
  3,
  2369,
  1550,
  2370,
  1,
  423,
  4,
  1],
 [1,
  155,
  21,
  2368,
  1549,
  8

Finding Maximum Sequence Length

 calculates the maximum sequence length among all input sequences generated from the text data.

In [ ]:
max_length = max([len(x) for x in input_sequences])
max_length

325

Padding Input Sequences

pads the input sequences to ensure uniform length. It uses the pad_sequences function from TensorFlow Keras to pad the sequences with zeros (pre-padding) up to the maximum sequence length (max_len)

In [ ]:
padded_input_sequences = pad_sequences(input_sequences, maxlen = max_length, padding = 'pre')
padded_input_sequences

array([[   0,    0,    0, ...,    0,    1,  155],
       [   0,    0,    0, ...,    1,  155,   21],
       [   0,    0,    0, ...,  155,   21, 2368],
       ...,
       [   0,    0,    0, ..., 2331,  290,   19],
       [   0,    0,    0, ...,  290,   19,   54],
       [   0,    0,    0, ...,   19,   54, 1535]], dtype=int32)

In [ ]:
X = padded_input_sequences[:,:-1]
X

array([[   0,    0,    0, ...,    0,    0,    1],
       [   0,    0,    0, ...,    0,    1,  155],
       [   0,    0,    0, ...,    1,  155,   21],
       ...,
       [   0,    0,    0, ...,   64, 2331,  290],
       [   0,    0,    0, ..., 2331,  290,   19],
       [   0,    0,    0, ...,  290,   19,   54]], dtype=int32)

In [ ]:
y = padded_input_sequences[:,-1]
y

array([ 155,   21, 2368, ...,   19,   54, 1535], dtype=int32)

In [ ]:
X.shape

(26383, 324)

In [ ]:
y.shape

(26383,)

One-Hot Encoding Labels

one-hot encoding on the labels (y) to convert them into categorical format. It utilizes the to_categorical function from TensorFlow Keras to encode the labels into binary vectors

In [ ]:
y = to_categorical(y,num_classes=len(tokenizer.word_index)+1)

In [ ]:
y.shape

(26383, 4994)

In [ ]:
y

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

LSTM Language Model Architecture -




1.) Embedding Layer: This layer converts integer indices into dense vectors of fixed size. It takes input sequences of length 325 nd outputs dense vectors of size 100 for each word in the input sequence.

2.) LSTM Layer: This layer consists of 150 LSTM units. It takes the output of the embedding layer as input and processes the sequential data, capturing dependencies among words in the input sequence.

3.)Dense Layer: This layer is a fully connected (dense) layer with softmax activation function. It takes the output of the LSTM layer and predicts the probability distribution over all words in the vocabulary

In [ ]:
## Define the LSTM language model architecture
model = Sequential()
model.add(Embedding(4994,100,input_length=324))
model.add(LSTM(150))
model.add(Dense(4994,activation='softmax'))


## Compile the model
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
model.build(input_shape=(324, 100))

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


In [ ]:
 ##Summary of the Model
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding (Embedding)                │ (324, 100, 100)             │         499,400 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ lstm (LSTM)                          │ (324, 150)                  │         150,600 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (324, 4994)                 │         754,094 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 1,404,094 (5.36 MB)

 Trainable params: 1,404,094 (5.36 MB)

 Non-trainable params: 0 (0.00 B)

Training our model

We will train  defined LSTM language model (model) using the input features (X) and one-hot encoded labels (y) for a specified number of epochs (150) .During training, the model learns to predict the next word in a sequence based on the input context provided by the input sequences (X).

In [ ]:
# Train the LSTM language model
model.fit(X, y, epochs=150)

Epoch 1/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 17s 17ms/step - accuracy: 0.0501 - loss: 7.2850
Epoch 2/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 19s 17ms/step - accuracy: 0.0765 - loss: 6.3409
Epoch 3/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 20s 17ms/step - accuracy: 0.0943 - loss: 5.8973
Epoch 4/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 20s 17ms/step - accuracy: 0.1219 - loss: 5.4759
Epoch 5/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 21s 17ms/step - accuracy: 0.1390 - loss: 5.1311
Epoch 6/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 14s 17ms/step - accuracy: 0.1588 - loss: 4.7589
Epoch 7/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 20s 17ms/step - accuracy: 0.1865 - loss: 4.4043
Epoch 8/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 20s 17ms/step - accuracy: 0.2132 - loss: 4.0968
Epoch 9/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 14s 17ms/step - accuracy: 0.2593 - loss: 3.7623
Epoch 10/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 14s 17ms/step - accuracy: 0.3063 - loss: 3.4504
Epoch 11/150
825/825 ━━━━━━━━━━━━━━━━━━━━ 14s 17ms/step - accuracy: 0.3582 - loss: 3.1680
Epoch 12/150
825/82

 Generating Text using Trained Language Model

In [ ]:
text = "Please let me know"

for i in range(25):
    # Tokenize the input text
    token_text = tokenizer.texts_to_sequences([text])[0]
    # Pad the tokenized text
    padded_token_text = pad_sequences([token_text], maxlen=max_length, padding='pre')
    # Predict the next word index
    pos = np.argmax(model.predict(padded_token_text))

    # Retrieve the word corresponding to the predicted index
    for word, index in tokenizer.word_index.items():
        if index == pos:
            # Append the predicted word to the input text
            text = text + " " + word
            print(text)
            # Simulate typing effect with a delay of 1 second
            time.sleep(1)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 39ms/step
Please let me know if
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step
Please let me know if you
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 56ms/step
Please let me know if you have
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step
Please let me know if you have any
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step
Please let me know if you have any other
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step
Please let me know if you have any other requests
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step
Please let me know if you have any other requests or
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step
Please let me know if you have any other requests or would
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step
Please let me know if you have any other requests or would like
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step
Please let me know if you have any other requests or would like me
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step
Please let me know if you have any other requests or would like me to
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 39ms/step
Please let me know if you 

Saving model

In [ ]:
import pickle

In [ ]:
with open('trained_model.pkl', 'wb') as file:
       pickle.dump(model, file)